linkerd/app/outbound/src/http/logical.rs

```diff
@@ -14,6 +14,267 @@ use tracing::debug_span;
+                // unify the profile stack's response type with that of to
+                // endpoint stack.
+                .push(http::NewHeaderFromTarget::<CanonicalDstHeader, _>::layer())
+                .instrument(|l: &Logical| debug_span!("logical", dst = %l.logical_addr))
+                .push_on_service(svc::BoxService::layer())
+                .push(svc::ArcNewService::layer())
+        })
+    }
+
+    /// Push the stack that's built when a client policy is discovered for an
+    /// outbound service.
+    fn push_client_policy<ESvc>(
+        self,
+    ) -> Outbound<
+        impl svc::NewService<
+                Logical,
+                Service = impl svc::Service<
+                    http::Request<http::BoxBody>,
+                    Response = http::Response<http::BoxBody>,
+                    Future = impl Send,
+                    Error = impl Into<Error>,
+                > + Send
+                              + 'static,
+            > + Clone,
+    >
+    where
+        E: svc::NewService<Concrete, Service = ESvc> + Clone + Send + Sync + 'static,
+        ESvc: svc::Service<http::Request<http::BoxBody>, Response = http::Response<http::BoxBody>>
+            + Send
+            + 'static,
+        ESvc::Error: Into<Error>,
+        ESvc::Future: Send,
+    {
+        self.map_stack(|config, rt, concrete| {
+            let config::ProxyConfig {
+                buffer_capacity,
+                dispatch_timeout,
+                ..
+            } = config.proxy;
+
+            concrete
+                .push_map_target(Concrete::from)
+                // Any per-route policy has to go between the split layer and
+                // the concrete stack cache, since we _don't_ want to share
+                // client policy across routes that have the same backend!
+                .push_map_target(
+                    |(concrete, PolicyRoute { route: _, logical }): (ConcreteAddr, PolicyRoute)| {
+                        (concrete, logical)
+                    },
+                )
+                .push(policy::split::layer())
+                .push_on_service(
+                    svc::layers()
+                        .push(svc::layer::mk(svc::SpawnReady::new))
+                        .push(
+                            rt.metrics
+                                .proxy
+                                .stack
+                                .layer(stack_labels("http", "HTTPRoute")),
+                        )
+                        .push(svc::FailFast::layer("HTTPRoute", dispatch_timeout))
+                        .push_spawn_buffer(buffer_capacity),
+                )
+                .check_new_clone::<PolicyRoute>()
+                .push_map_target(|(route, logical)| {
+                    tracing::debug!(?route);
+                    PolicyRoute { route, logical }
+                })
+                .push(policy::http::NewServiceRouter::layer())
```

Comment:
If the concrete stack is cached, can we omit the cache/buffer around the split so that every route gets its own split?

If so, perhaps it makes sense to pull it all into one layer that is Logical-outside and Concrete-inside... We'll need to consider how/where per-route functionality is added for things like retries/tmeouts.